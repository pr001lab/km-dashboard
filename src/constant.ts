// export const API = 'http://localhost:3100';

export const API =
  process.env.NODE_ENV === 'development'
    ? ``
    : `https://raw.githubusercontent.com/pr001lab/km-dashboard/gh-pages/`;

export enum LoadingStatus {
  Idle = 'idle',
  Loading = 'loading',
  Successed = 'successed',
  Failed = 'failed',
}

export enum APIRoute {
  Sites = 'sites',
  Tests = 'tests',
  SitesJSONFile = 'sites.json',
  TestsJSONFile = 'tests.json',
}

export enum AppRoute {
  Main = '/',
  ResultsTestId = '/results',
  FinalizeTestId = '/finalize',
}

export enum TitlePageAppRoute {
  Main = 'Dashboard',
  Results = 'Results',
  Finalize = 'Finalize',
  Error = '404',
}

export enum Type {
  CLASSIC = 'CLASSIC',
  SERVER_SIDE = 'SERVER_SIDE',
  MVT = 'MVT',
}

export enum Status {
  DRAFT = 'DRAFT',
  ONLINE = 'ONLINE',
  PAUSED = 'PAUSED',
  STOPPED = 'STOPPED',
}

export enum SortEnum {
  Ascending = 'Ascending',
  Descending = 'Descending',
  None = 'None',
}

export enum fieldSort {
  STATUS = 'status',
}

export const tableColumns = ['name', 'type', 'status', 'site'];

export const orderSortStatus = ['ONLINE', 'PAUSED', 'STOPPED', 'DRAFT'];
