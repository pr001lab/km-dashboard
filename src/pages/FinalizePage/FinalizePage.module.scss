@import '../../styles/_vars.scss';

.wrapper {
  flex-grow: 1;
}

.table {
  font-family: $defaultFontFamily;
  display: flex;
  flex-direction: column;

  & thead {
    text-align: left;
    color: #999;
    font-size: 11px;
    font-weight: 500;
    line-height: normal;
    text-transform: uppercase;

    & .table-header {
      padding: 10px;
      width: unset;
      height: unset;
      background: none;
    }
  }

  & tr {
    width: 100%;
    display: grid;
    grid-template-columns: 4px 1fr 115px 106px 167px 95px;
    align-items: center;
    gap: 10px;
    padding-right: 12px;
  }

  & tbody {
    width: 100%;
    display: flex;
    flex-direction: column;
    row-gap: 5px;
  }

  & tbody tr {
    color: #5c5c5c;
    font-size: 12px;
    line-height: normal;
    font-weight: 500;
    min-height: 70px;
    border-radius: 3px;
    background: #fff;

    & .row-indicator {
      height: 100%;
      width: 4px;
      border-radius: 3px 0 0 3px;

      &.siteId-1 {
        background: #e14165;
      }

      &.siteId-2 {
        background: #c2c2ff;
      }

      &.siteId-3 {
        background: #8686ff;
      }
    }

    & .row-name {
      color: #474747;
      font-size: 13px;
    }

    & .row-type {
      font-weight: 400;
    }

    & .row-status {
      &.ONLINE {
        color: #1bda9d;
      }

      &.PAUSED,
      &.STOPPED {
        color: #ff8346;
      }
    }

    & .row-site {
      font-weight: 400;
    }

    & td {
      padding: 15px 0;
    }
  }

  & tbody tr:hover {
    background: #fff;
    box-shadow: 0 0 20px 0 rgba(0, 0, 0, 0.15);
  }
}

.link-back {
  display: flex;
  align-items: center;
  gap: 5px;
}

.link-back span {
  color: $text-heading;
  font-family: $linkFontFamily;
  font-size: 18px;
  font-style: normal;
  font-weight: 700;
  line-height: normal;
}
