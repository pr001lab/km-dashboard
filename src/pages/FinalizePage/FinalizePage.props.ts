import { ComponentProps } from 'react';

export interface FinalizePageProps extends ComponentProps<'div'> {}
