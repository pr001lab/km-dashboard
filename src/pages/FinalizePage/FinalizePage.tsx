import Heading from '../../components/common/Heading/Heading';
import styles from '../ResultsPage/ResultsPage.module.scss';
import Footer from '../../components/common/Footer/Footer';
import { Link, useNavigate, useParams } from 'react-router-dom';
import { ReactComponent as Chevron } from '../../assets/icons/Chevron.svg';
import Loader from '../../components/common/Loader/Loader';
import { useEffect, useState } from 'react';
import { Site, Test } from '../../types';
import axios from 'axios';
import { API, APIRoute, AppRoute, Type } from '../../constant';
import Wrapper from '../../components/common/Wrapper/Wrapper';
import { getTextTransformCapitalize } from '../../utils';
import LinkButton from '../../components/common/LinkButton/LinkButton';

function FinalizePage() {
  const navigate = useNavigate();
  const { testId } = useParams();
  const [loading, setLoading] = useState<boolean>(false);
  const [tests, setTests] = useState<Test[]>([]);

  useEffect(() => {
    setLoading(true);
    Promise.all([getSites(), getTests()])
      .then(([sitesResponse, testsResponse]) => {
        const tests = testsResponse.map((test) => {
          const site = sitesResponse.find((site) => site.id === test.siteId);
          test.siteUrl =
            site?.url && new URL(site?.url).hostname.replace(/^www\./, '');
          return test;
        });

        setTests(tests);
        setLoading(false);
      })
      .catch((error) => {
        console.error('Error fetching data:', error);
        setLoading(false);
      });
  }, []);

  const getSites = async () => {
    const { data } = await axios.get<Site[]>(`${API}/${APIRoute.Sites}`);
    return data;
  };

  const getTests = async () => {
    const { data } = await axios.get<Test[]>(
      `${API}/${APIRoute.Tests}/${testId}`,
    );
    return Array.isArray(data) ? data : [data];
  };

  if (loading) {
    return <Loader />;
  }

  return (
    <>
      <Wrapper className={styles['wrapper']}>
        <Heading level="h1">Finalize</Heading>
        <Heading level="h2">Spring promotion</Heading>
        <table className={styles['table']}>
          <thead>
            <tr>
              <th></th>
              <th>
                <Wrapper className={styles['table-header']}>Name</Wrapper>
              </th>
              <th>
                <Wrapper className={styles['table-header']}>Type</Wrapper>
              </th>
              <th>
                <Wrapper className={styles['table-header']}>Status</Wrapper>
              </th>
              <th>
                <Wrapper className={styles['table-header']}>Site</Wrapper>
              </th>
              <th></th>
            </tr>
          </thead>
          <tbody>
            {tests
              .slice()
              .sort()
              .map((test) => {
                const { id, name, type, status, siteId, siteUrl } = test;
                const indicatorColor = `siteId-${siteId}`;
                const typeTransform = [Type.MVT].includes(type)
                  ? type
                  : getTextTransformCapitalize(type).replace(/[_]/g, '-');
                const linkTo =
                  status === 'DRAFT'
                    ? `${AppRoute.FinalizeTestId}/${id}`
                    : `${AppRoute.ResultsTestId}/${id}`;

                return (
                  <tr className={styles['row-table']} key={id}>
                    <td
                      className={`${styles['row-indicator']} ${styles[indicatorColor]}`}
                    ></td>
                    <td className={styles['row-name']}>{name}</td>
                    <td className={styles['row-type']}>{typeTransform}</td>
                    <td className={`${styles['row-status']} ${styles[status]}`}>
                      {getTextTransformCapitalize(status)}
                    </td>
                    <td className={styles['row-site']}>{siteUrl}</td>
                    <td>
                      <LinkButton
                        to={linkTo}
                        className={status === 'DRAFT' ? 'gray-dark' : ''}
                      >
                        {status === 'DRAFT' ? 'Finalize' : 'Results'}
                      </LinkButton>
                    </td>
                  </tr>
                );
              })}
          </tbody>
        </table>
      </Wrapper>
      <Footer>
        <Link
          className={styles['link-back']}
          onClick={() => navigate(-1)}
          to={'#'}
        >
          <Chevron />
          <span className={styles['link-label']}>Back</span>
        </Link>
      </Footer>
    </>
  );
}

export default FinalizePage;
