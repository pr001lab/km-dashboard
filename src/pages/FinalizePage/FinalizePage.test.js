import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { TitlePageAppRoute } from '../../constant';
import FinalizePage from './FinalizePage';

const mockDataTests = {
  tests: [
    {
      'id': 2,
      'name': 'Dark theme test',
      'type': 'MVT',
      'status': 'DRAFT',
      'siteId': 3,
    },
  ],
};

jest.mock('../../hooks/useGetData', () => ({
  useGetData: () => mockDataTests,
}));

describe('FinalizePage tests', () => {
  it('should render FinalizePage for the path with testId', async () => {
    const view = render(
      <MemoryRouter
        future={{
          v7_relativeSplatPath: true,
          v7_startTransition: true,
        }}
      >
        <FinalizePage />
      </MemoryRouter>,
    );
    screen.getByRole('heading', { name: TitlePageAppRoute.Finalize });
    expect(view).toMatchSnapshot();
  });
});
