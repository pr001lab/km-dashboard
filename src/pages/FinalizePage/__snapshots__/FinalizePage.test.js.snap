// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinalizePage tests should render FinalizePage for the path with testId 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="wrapper wrapper"
      >
        <h1
          class="heading"
        >
          Finalize
        </h1>
        <h2
          class="heading"
        >
          Spring promotion
        </h2>
        <table
          class="table"
        >
          <thead>
            <tr>
              <th />
              <th>
                <div
                  class="wrapper table-header"
                >
                  Name
                </div>
              </th>
              <th>
                <div
                  class="wrapper table-header"
                >
                  Type
                </div>
              </th>
              <th>
                <div
                  class="wrapper table-header"
                >
                  Status
                </div>
              </th>
              <th>
                <div
                  class="wrapper table-header"
                >
                  Site
                </div>
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              class="row-table"
            >
              <td
                class="thead th1"
              >
                Name
              </td>
              <td
                class="thead th2"
              >
                Type
              </td>
              <td
                class="thead th3"
              >
                Status
              </td>
              <td
                class="thead th4"
              >
                Site
              </td>
              <td
                class="row-indicator siteId-3"
              />
              <td
                class="row-name"
              >
                Dark theme test
              </td>
              <td
                class="row-type"
              >
                MVT
              </td>
              <td
                class="row-status DRAFT"
              >
                Draft
              </td>
              <td
                class="row-site"
              />
              <td
                class="row-actions"
              >
                <a
                  class="link-button gray-dark"
                  href="/finalize/2"
                >
                  Finalize
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <footer
        class="footer"
      >
        <a
          class="link-back"
          href="/"
        >
          <svg>
            Chevron.svg
          </svg>
          <span
            class="link-label"
          >
            Back
          </span>
        </a>
      </footer>
    </div>
  </body>,
  "container": <div>
    <div
      class="wrapper wrapper"
    >
      <h1
        class="heading"
      >
        Finalize
      </h1>
      <h2
        class="heading"
      >
        Spring promotion
      </h2>
      <table
        class="table"
      >
        <thead>
          <tr>
            <th />
            <th>
              <div
                class="wrapper table-header"
              >
                Name
              </div>
            </th>
            <th>
              <div
                class="wrapper table-header"
              >
                Type
              </div>
            </th>
            <th>
              <div
                class="wrapper table-header"
              >
                Status
              </div>
            </th>
            <th>
              <div
                class="wrapper table-header"
              >
                Site
              </div>
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr
            class="row-table"
          >
            <td
              class="thead th1"
            >
              Name
            </td>
            <td
              class="thead th2"
            >
              Type
            </td>
            <td
              class="thead th3"
            >
              Status
            </td>
            <td
              class="thead th4"
            >
              Site
            </td>
            <td
              class="row-indicator siteId-3"
            />
            <td
              class="row-name"
            >
              Dark theme test
            </td>
            <td
              class="row-type"
            >
              MVT
            </td>
            <td
              class="row-status DRAFT"
            >
              Draft
            </td>
            <td
              class="row-site"
            />
            <td
              class="row-actions"
            >
              <a
                class="link-button gray-dark"
                href="/finalize/2"
              >
                Finalize
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <footer
      class="footer"
    >
      <a
        class="link-back"
        href="/"
      >
        <svg>
          Chevron.svg
        </svg>
        <span
          class="link-label"
        >
          Back
        </span>
      </a>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
